packaging/Makefile

```diff
@@ -0,0 +1,120 @@
+PLATFORM = freebsd
+else ifeq ($(UNAME_S),NetBSD)
+PLATFORM = netbsd
+else ifeq ($(UNAME_S),OpenBSD)
+PLATFORM = openbsd
+else ifeq ($(UNAME_S),DragonFly)
+PLATFORM = dragonfly
+else ifeq ($(shell uname -o),Msys)
+PLATFORM = msys2
+else
+$(error Unable to detect platform. Please open a ticket with the output of uname -a.)
+endif
+endif
+
+all: packages
+	@:
+
+# --------------------------------------------------------------------
+# Packaging.
+# --------------------------------------------------------------------
+
+.PHONY: packages package-deb \
+	package-rpm \
+	package-rpm-redhat package-rpm-fedora \
+	package-rpm-rhel6 package-rpm-rhel7 package-rpm-rhel8 \
+	package-rpm-suse package-rpm-opensuse package-rpm-sles11 \
+	package-windows \
+	package-generic-unix \
+	docker-image
+
+PACKAGES_DIR ?= ../PACKAGES
+SOURCE_DIST_FILE ?= $(wildcard $(PACKAGES_DIR)/rabbitmq-server-*.tar.xz)
+
+ifneq ($(filter-out clean,$(MAKECMDGOALS)),)
+ifeq ($(SOURCE_DIST_FILE),)
+$(error Cannot find source archive; please specify SOURCE_DIST_FILE)
+endif
+ifneq ($(words $(SOURCE_DIST_FILE)),1)
+$(error Multiple source archives found; please specify SOURCE_DIST_FILE)
+endif
+ifeq ($(filter %.tar.xz %.txz,$(SOURCE_DIST_FILE)),)
+$(error The source archive must a tar.xz archive)
+endif
+ifeq ($(wildcard $(SOURCE_DIST_FILE)),)
+$(error The source archive must exist)
+endif
+endif
+
+ifndef NO_CLEAN
+DO_CLEAN := clean
+endif
+
+VARS = SOURCE_DIST_FILE="$(abspath $(SOURCE_DIST_FILE))" \
+       PACKAGES_DIR="$(abspath $(PACKAGES_DIR))" \
+       SIGNING_KEY="$(SIGNING_KEY)"
+
+packages: package-deb package-rpm package-windows package-generic-unix
+	@:
+
+package-deb: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C debs/Debian $(VARS) all $(DO_CLEAN)
+
+package-rpm: package-rpm-redhat package-rpm-suse
+	@:
+
+# FIXME: Why not package-rpm-fedora?
+package-rpm-redhat: package-rpm-rhel6 package-rpm-rhel7 package-rpm-rhel8
+	@:
+
+package-rpm-fedora: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C RPMS/Fedora $(VARS) all $(DO_CLEAN)
+
+package-rpm-rhel6: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C RPMS/Fedora $(VARS) RPM_OS=rhel6 all $(DO_CLEAN)
+
+package-rpm-rhel7: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C RPMS/Fedora $(VARS) RPM_OS=rhel7 all $(DO_CLEAN)
+
+package-rpm-rhel8: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C RPMS/Fedora $(VARS) RPM_OS=rhel8 all $(DO_CLEAN)
+
+package-rpm-suse: package-rpm-opensuse package-rpm-sles11
+	@:
+
+package-rpm-opensuse: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C RPMS/Fedora $(VARS) RPM_OS=opensuse all $(DO_CLEAN)
+
+package-rpm-sles11: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C RPMS/Fedora $(VARS) RPM_OS=sles11 all $(DO_CLEAN)
+
+package-windows: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C windows $(VARS) all $(DO_CLEAN)
+	$(verbose) $(MAKE) -C windows-exe $(VARS) all $(DO_CLEAN)
+
+package-generic-unix: $(SOURCE_DIST_FILE)
+	$(gen_verbose) $(MAKE) -C generic-unix $(VARS) all $(DO_CLEAN)
+
+docker-image:
```

Comment:
`make docker-image` doesn't work because of the RabbitMQ version format. 

ex:
```bash
$  make docker-image
.....
docker build --pull \
          --build-arg PGP_KEYSERVER=pgpkeys.uk \
          --build-arg OTP_VERSION=23.2.4 \
          --build-arg OTP_SHA256=e72aa084907e0f34f932cf00caa33aba93147b0a7c9c35569d6bd1c402f532de \
          --build-arg RABBITMQ_VERSION=3.8.10+66.gee45535 \
          --build-arg RABBITMQ_BUILD=rabbitmq_server-3.8.10+66.gee45535 \
          --tag pivotalrabbitmq/rabbitmq:3.8.10+66.gee45535 \
          .
invalid argument "pivotalrabbitmq/rabbitmq:3.8.10+66.gee45535" for "-t, --tag" flag: invalid reference format
See 'docker build --help'.
make[2]: *** [dist] Error 125
make[1]: *** [docker-image] Error 2
make: *** [docker-image] Error 2
```

The docker `tag` doesn't support `+` (ex `--tag pivotalrabbitmq/rabbitmq:3.8.10+66.gee45535 \`)  so we should replace `+` with something else when we tag the image

Reply:
This is what I did last time I had to run this:

```
make source-dist PROJECT_VERSION=3.8.12-rc.3-20200212.1331
make package-generic-unix SOURCE_DIST_FILE=/Users/gerhard/github.com/rabbitmq/rabbitmq-server/PACKAGES/rabbitmq-server-3.8.12-rc.3-20200212.1331.tar.xz
make docker-image SOURCE_DIST_FILE=/Users/gerhard/github.com/rabbitmq/PACKAGES/rabbitmq-server-generic-unix-3.8.12-rc.3-20200212.1331.tar.xz
```

Reply:
Depending on the circumstances, I have found that using a date as above can be more convenient when iterating on unpublished changes. However, I doubt there is an ideal solution for all cases. It's also worth noting that if you have uncommitted changes, you will currently get a version such as `rabbitmq-server-3.8.0+rc.1.733.gee45535.dirty` which is reasonable, but not reproducible. I think replacing `+` with `-` at the point of tagging would be reasonable, since I think it's important that setting `PROJECT_VERSION` as a make var still has the intended effect.